Optimize PublicSuffixStore

cdumez opened the pull request:
#### 1becc19b6e5ecf20c63b152e5cd5591e1ab9566d
<pre>
Optimize PublicSuffixStore
<a href="https://bugs.webkit.org/show_bug.cgi?id=274358">https://bugs.webkit.org/show_bug.cgi?id=274358</a>

Reviewed by Darin Adler and Sam Weinig.

* Source/WTF/wtf/HashMap.h:
(WTF::HashMapEnsureTranslatorAdapter::hash):
(WTF::HashMapEnsureTranslatorAdapter::equal):
(WTF::HashMapEnsureTranslatorAdapter::translate):
(WTF::TableTraitsArg&gt;::ensure):
* Source/WTF/wtf/URLHelpers.h:
* Source/WTF/wtf/text/StringHash.h:
(WTF::ASCIICaseInsensitiveStringViewHashTranslator::translate):
* Source/WTF/wtf/text/StringImpl.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::urlForBindings const):
* Source/WebCore/page/Quirks.cpp:
(WebCore::isYahooMail):
(WebCore::Quirks::shouldHideSearchFieldResultsButton const):
(WebCore::Quirks::isAmazon const):
(WebCore::Quirks::isGoogleMaps const):
* Source/WebCore/platform/PublicSuffixStore.cpp:
(WebCore::PublicSuffixStore::topPrivatelyControlledDomain const):
* Source/WebCore/platform/PublicSuffixStore.h:
(WebCore::PublicSuffixStore::WTF_GUARDED_BY_LOCK): Deleted.
* Source/WebCore/platform/cocoa/PublicSuffixStoreCocoa.mm:
(WebCore::isPublicSuffixCF):
(WebCore::PublicSuffixStore::platformIsPublicSuffix const):
(WebCore::PublicSuffixStore::platformTopPrivatelyControlledDomain const):
(WebCore::PublicSuffixStore::enablePublicSuffixCache):
(WebCore::PublicSuffixStore::addPublicSuffix):
* Source/WebCore/platform/network/curl/PublicSuffixStoreCurl.cpp:
(WebCore::PublicSuffixStore::platformTopPrivatelyControlledDomain const):
* Source/WebCore/platform/soup/PublicSuffixStoreSoup.cpp:
(WebCore::PublicSuffixStore::platformTopPrivatelyControlledDomain const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::resetStateAfterProcessTermination):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::processDidTerminateOrFailedToLaunch):
(WebKit::WebProcessProxy::didExceedMemoryFootprintThreshold):

Canonical link: <a href="https://commits.webkit.org/278977@main">https://commits.webkit.org/278977@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/30753a7979caf9b1f871cf4edd46373ff5811ab3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52154 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31485 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4574 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55427 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2868 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54458 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37884 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2575 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42442 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/1839 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54250 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29120 "Found 1 new test failure: http/tests/site-isolation/draw-after-cross-origin-navigation-between-two-remote-frames.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45010 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23513 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/51998 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26388 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2284 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1036 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/45503 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48293 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2430 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57024 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/51662 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27279 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/2485 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49840 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28513 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45121 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49073 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29425 "Built successfully") | | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/63969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7624 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28257 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/63969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/HashMap.h:
```diff
@@ -187,6 +187,7 @@ class HashMap final {
     //   static bool equal(const ValueType&, const T&);
     //   static translate(ValueType&, const T&, unsigned hashCode);
     template<typename HashTranslator, typename K, typename V> AddResult add(K&&, V&&);
+    template<typename HashTranslator, typename K, typename Functor> AddResult ensure(K&&, Functor&&);
```
Comments above mention only `find` and `add`, yet we have `contains`, `get`, `inlineGet`, `remove`, and `ensure`.

darinadler commented on Source/WebCore/platform/PublicSuffixStore.cpp:
```diff
@@ -71,30 +72,32 @@ String PublicSuffixStore::publicSuffix(const URL& url) const
 
     if (!host.containsOnlyASCII())
-        return host;
+        return host.toString();
```
Aren’t we losing an optimization here where we could reuse a passed-in string?

cdumez replied:
True but:
1. More than half of the call sites have a StringView, not a String
2. This is the case where the URL host contains non-ASCII characters, I don't think this is a case we need to optimize for.

darinadler reviewed (approved):
Did you measure a speed up?

weinig commented on Source/WebCore/platform/cocoa/PublicSuffixStoreCocoa.mm:
```diff
@@ -27,63 +27,64 @@
     {
         Locker locker { m_publicSuffixCacheLock };
-        if (m_publicSuffixCache && m_publicSuffixCache->contains(domainString))
+        if (m_publicSuffixCache && m_publicSuffixCache->contains<ASCIICaseInsensitiveStringViewHashTranslator>(domain))
```
Based on the assumption that anyone who uses `HashSet<String, ASCIICaseInsensitiveHash>` would want this translation if they happened to pass in a StringView, I think we should probably take the time to build the infrastructure to make this automatic without needing to pass the translator explicitly.

One approach would be to extend the Hash argument to support exposing a set of "related" or "translatable" types and their associated translators. Then StringHash (and ASCIICaseInsensitiveHash, etc.) would then provide the means for `String`, `StringView`, `ASCIILiteral`, etc to all be used in the translatable functions (contains(), add(), etc.). This would probably also provide a way to replace the "smart pointer" specific overloads I added a long time ago, with a more generic replacement.

Very roughly, the has containers would get new functions that look something like:

```c++
    template<typename RelatedType> 
        requires IsRelatedType<HashArg , RelatedType>  
    iterator find(const RelatedType& key) const
    {
        return find<LookupTranslatorForRelatedType<HashArg, RelatedType>>(std::forward<RelatedType>(key)); 
    }
 ```

and StringHash, for example, would get something akin to:

```c++
    struct StringHash {
        ...
        using Supported = TypeList<
             Alt<String, StringTranslator>, 
             Alt<StringImpl*, StringImplTranslator>, 
             Alt<StringView, StringViewTranslator>, 
             Alt<ASCIILiteral, ASCIILiteralTranslator>
        >;
    };
```

(plenty of other ways to structure that, such as using overloading, but you get the idea).

weinig commented on Source/WebCore/platform/PublicSuffixStore.h:
```diff
@@ -53,14 +53,14 @@ class PublicSuffixStore {
     mutable Lock m_publicSuffixCacheLock;
     std::optional<HashSet<String, ASCIICaseInsensitiveHash>> m_publicSuffixCache WTF_GUARDED_BY_LOCK(m_publicSuffixCacheLock);
-    bool m_canAcceptCustomPublicSuffix WTF_GUARDED_BY_LOCK(m_publicSuffixCacheLock) { false };
+    bool m_canAcceptCustomPublicSuffix { false }; // Only used on the main thread.
```
Obviously not this for this PR, but might be worth investing in a wrapper than can at least assert main thread only for debug builds.

weinig commented on Source/WebCore/platform/PublicSuffixStore.cpp:
```diff
@@ -71,30 +72,32 @@ String PublicSuffixStore::publicSuffix(const URL& url) const
-        m_hostTopPrivatelyControlledDomainCache.remove(m_hostTopPrivatelyControlledDomainCache.random());
+    auto domain = addResult.iterator->value.isolatedCopy();
+    if (!addResult.isNewEntry) {
+        constexpr auto maxHostTopPrivatelyControlledDomainCache = 128;
```
I'd probably put this constant up at the top of the file so it can be more easily seen and tweaked if needed (I realized you are just moving it a bit).

weinig reviewed (approved):
A bit more depth in the changelog would be good.

cdumez committed:
Optimize PublicSuffixStore
https://bugs.webkit.org/show_bug.cgi?id=274358

Reviewed by Darin Adler and Sam Weinig.

* Source/WTF/wtf/HashMap.h:
(WTF::HashMapEnsureTranslatorAdapter::hash):
(WTF::HashMapEnsureTranslatorAdapter::equal):
(WTF::HashMapEnsureTranslatorAdapter::translate):
(WTF::TableTraitsArg>::ensure):
* Source/WTF/wtf/URLHelpers.h:
* Source/WTF/wtf/text/StringHash.h:
(WTF::ASCIICaseInsensitiveStringViewHashTranslator::translate):
* Source/WTF/wtf/text/StringImpl.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::urlForBindings const):
* Source/WebCore/page/Quirks.cpp:
(WebCore::isYahooMail):
(WebCore::Quirks::shouldHideSearchFieldResultsButton const):
(WebCore::Quirks::isAmazon const):
(WebCore::Quirks::isGoogleMaps const):
* Source/WebCore/platform/PublicSuffixStore.cpp:
(WebCore::PublicSuffixStore::topPrivatelyControlledDomain const):
* Source/WebCore/platform/PublicSuffixStore.h:
(WebCore::PublicSuffixStore::WTF_GUARDED_BY_LOCK): Deleted.
* Source/WebCore/platform/cocoa/PublicSuffixStoreCocoa.mm:
(WebCore::isPublicSuffixCF):
(WebCore::PublicSuffixStore::platformIsPublicSuffix const):
(WebCore::PublicSuffixStore::platformTopPrivatelyControlledDomain const):
(WebCore::PublicSuffixStore::enablePublicSuffixCache):
(WebCore::PublicSuffixStore::addPublicSuffix):
* Source/WebCore/platform/network/curl/PublicSuffixStoreCurl.cpp:
(WebCore::PublicSuffixStore::platformTopPrivatelyControlledDomain const):
* Source/WebCore/platform/soup/PublicSuffixStoreSoup.cpp:
(WebCore::PublicSuffixStore::platformTopPrivatelyControlledDomain const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::resetStateAfterProcessTermination):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::processDidTerminateOrFailedToLaunch):
(WebKit::WebProcessProxy::didExceedMemoryFootprintThreshold):

Canonical link: https://commits.webkit.org/278977@main
